doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,321 @@
+repository, an HTTP endpoint you define will get called so you can receive the event in near
+real-time.
+
+Supporting webhooks in Materialize allows users to easily get their data into MZ, and the low
+latency of the events works very well with the premise of being an "operational" data warehouse.
+
+> **_Note_** To achieve the goal of "make it easier to get data in Materialize" we also considered
+building support for Fivetran. We may still do this in the future, but the amount of work to get
+webhooks working is significantly less, and the frequency of data is more suited to the goals of
+Materialize.
+
+## Goals
+
+1. Provide URLs that users can push data to.
+2. Support an extensible syntax similar to `CREATE SOURCE <name> FROM WEBHOOK` that allows
+users to generate the aforementioned URLs.
+3. Handle at least a total of 100 requests per-second across all `WEBHOOK` sources.
+5. Code complete by August 15th, 2023.
+
+
+## Non-Goals
+
+The following are all features we want to support, but are out of scope in the immediate term.
+
+* Extracting keys from the data for `UPSERT`-like behavior.
+* Custom filtering of headers.
+* Manipulating the incoming data, splitting a single request into any kind of "sub-sources".
+* Handling batch requests.
+
+## Overview
+
+Our implementation will mostly exist in the adapter, and we can re-use existing APIs on the
+`storage-client` to handle the persistence of data.
+
+We'll update the SQL parser to support our new kind of `WEBHOOK` source, and add a new endpoint to
+`environmentd`'s HTTP server `/api/webhook/:database/:schema/:name`. When a user creates this new
+kind of source, we'll persist the new `Source` in the catalog, use the `storage-client` to create a
+new collection.
+
+When we receive a event we'll use the `mz_adapter::Client` that the HTTP server has to send the raw
+`headers` and `body` to the coordinator, pack the data into a row, and append it to the necessary
+collection using a new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `WEBHOOK`, for which you can
+specify a `BODY FORMAT`, this denotes how we decode the body of a request. To start we'll support
+three formats, `BYTES`, `JSON`, or `TEXT`. By default we will not include the headers of the
+request, they can optionally be added by specifying `INCLUDE HEADERS`.
+
+```
+CREATE SOURCE <name> FROM WEBHOOK
+  BODY FORMAT [BYTES | JSON | TEXT]
+  (INCLUDE HEADERS)?
+;
+```
+
+After executing this request, we'll create a source with the following columns:
+
+| name    | type                        | optional?                                      |
+|---------|-----------------------------|------------------------------------------------|
+| body    | `bytea`, `jsonb`, or `text` | No                                             |
+| headers | `map[text -> text]`         | Yes, present if `INCLUDE HEADERS` is specified |
```

Comment:
Can HTTP headers be set more than once in a request for the same name? What about differently cased names?

Reply:
The spec says they can, but are case insensitive, I believe.

Reply:
I added an open question about the specific handling of HTTP headers. Because headers can be set more than once in a request, we could change the type of headers to be `map[text -> text[]]`, I'm not sure if this is supported in Materialize though?

Reply:
It is a supported type, as we discussed today, but also I think it's such an edge case that we should not care. At least, having the same header multiple times seems about as likely as having non-UTF-8 headers. If we were willing to punt on the latter, I think we can punt on the former, too, in the interest of the API being far more ergonomic for 99.9% of cases. We can just document that we only present the last header.

Reply:
Opened https://github.com/MaterializeInc/materialize/issues/20224 to finalize on a column type for `headers